### DISCUSSION TITLE
Redirect to 404

### DISCUSSION THREAD

**Discussion by stormfar** [vercel/next.js#10960] [Help] [answered]
How do you redirect to the 404 page from within `getStaticProps()`?

**Comment by timneutkens** [comment, marked as answer]
**Update:**
Since Next.js v10, you can return `notFound: true ` from `getStaticProps` which allows the page to return a 404 status and [404 Page](https://nextjs.org/docs/advanced-features/custom-error-page#404-page). Learn more here: https://nextjs.org/docs/api-reference/data-fetching/get-static-props#notfound

```js
export async function getStaticProps(context) {
  const res = await fetch(`https://.../data`)
  const data = await res.json()

  if (!data) {
    return {
      notFound: true,
    }
  }

  return {
    props: { data }, // will be passed to the page component as props
  }
}
```

---
I'm assuming in your case you're asking this because you're using the `fallback: true` behavior in getStaticPaths?

404 is a state of the UI so you should render it based on the properties:

```
// pages/blog/[slug].js

import {useRouter} from 'next/router'
import DefaultErrorPage from 'next/error'

export async function getStaticProps({ params }) {
  // fetch data from CMS through params
  const post = await getBlogItem(params.slug)
  return {
    props: {
      post
    }
  }
}

export async function getStaticPaths() {
  return {
    fallback: true,
    paths: []
  }
}

export default function MyPage({post}) {
  const router = useRouter()

  if(router.isFallback) {
     return <h1>Loading...</h1>
  }

  // This includes setting the noindex header because static files always return a status 200 but the rendered not found page page should obviously not be indexed
  if(!post) {
    return <>
      <Head>
        <meta name="robots" content="noindex">
      </Head>
      <DefaultErrorPage statusCode={404} />
    </>
  }

  return <h1>{post.title}</h1>
}
```

**Reply by stormfar** [reply]
Indeed. Ah, interesting. That's kind of what I was thinking but wasn't exactly sure. Thanks for the clarification!

**Reply by timneutkens** [reply]
Happy to help!

**Reply by qw-in** [reply]
@timneutkens thanks for this!

One quick question:
 If the dynamic route is called a second time with the same slug, I assume it will return the previously generated error page. Would there be a way to have it fallback again instead?

Example scenario:
1. User accidentally hits `/blog/my-post`
2. User correctly sees 404
3. I publish "my post" to `/blog/my-post`

**Reply by brianlovin** [reply]
That sounds like a good use case for the upcoming `revalidate` tag, although in the case of a new blog post, it might be easier and more resource-efficient to just redeploy 🙃

**Reply by tafelito** [reply]
@timneutkens is there a way of redirecting with a 404 status code instead of returning the error page with the error? is there any con of responding with status 200 instead of 404 even tho is not found?

**Reply by kristojorg** [reply]
@tafelito I also have this question. Would love to return a real 404 status code when my users hit a 404. To make that happen I wonder if I can redirect to a custom 404 page and then use `getServerSideProps` there to set the `res.status_code` to 404

**Reply by arunoda** [reply]
Thank @timneutkens 
This is gold.

**Reply by AshConnolly** [reply]
Thanks for the example @timneutkens! 😃 
Unfortunately this **doesn't seem to work if an error occurs** in `getStaticProps()` at any depth. Like if an error occurred in your get `getBlogItem` because it was trying to fetch data for a nonexistent page.

*Note: I arrived at this discussion from @Timer's comment https://github.com/vercel/next.js/discussions/11862 in the context of handling errors in `getStaticProps()`*

[Codesandbox example using your code](https://codesandbox.io/s/nextjs-getstaicprops-error-handling-k70y7?file=/pages/blog/%5Bslug%5D.js) - see lines 10 & 13

I'd want / **expect** to see the JSX block in `if(!post) {` showing the page-specific error
instead we **actually** get the global next app error page

Nautrally `getStaticProps()` might contain multiple methods which contain `try/catch` and might encounter errors.

Is there some way of showing page-based errors (instead of the global app error page), like in your example, for errors thrown in `getStaticProps()`? Apologies if I've misunderstood! 😃
[Tweet here also](https://twitter.com/_AshConnolly/status/1292481791769612288) if that offers more context.

**Reply by Vince-vegas** [reply]
This is what I've been looking for

on how to redirect to 404 page when getInitialProps's data has not been fetched the data or no data on it

**Reply by procarrera** [reply]
Hi guys,
Any news how to redirect user to 404 page if an error occur during getStaticProps? e.g user trying to access a non-existing page...?

thanks

**Reply by Pattygeek** [reply]
> I'm assuming in your case you're asking this because you're using the `fallback: true` behavior in getStaticPaths?
> 
> 404 is a state of the UI so you should render it based on the properties:
> 
> ```
> // pages/blog/[slug].js
> 
> import {useRouter} from 'next/router'
> import DefaultErrorPage from 'next/error'
> 
> export async function getStaticProps({ params }) {
>   // fetch data from CMS through params
>   const post = await getBlogItem(params.slug)
>   return {
>     props: {
>       post
>     }
>   }
> }
> 
> export async function getStaticPaths() {
>   return {
>     fallback: true,
>     paths: []
>   }
> }
> 
> export default function MyPage({post}) {
>   const router = useRouter()
> 
>   if(router.isFallback) {
>      return <h1>Loading...</h1>
>   }
> 
>   // This includes setting the noindex header because static files always return a status 200 but the rendered not found page page should obviously not be indexed
>   if(!post) {
>     return <>
>       <Head>
>         <meta name="robots" content="noindex">
>       </Head>
>       <DefaultErrorPage statusCode={404} />
>     </>
>   }
> 
>   return <h1>{post.title}</h1>
> }
> ```

I tried this in my code, but it is still not serving the error page.

**Reply by aguilarm** [reply]
Noting that in Next v10 you can return a 404 status code now https://nextjs.org/blog/next-10#notfound-support. Huzzah!

**Reply by parham-ab** [reply]
what if we use axios ?

**Comment by stevez86** [comment]
@timneutkens I'm a little concerned with returning a 200 success code for my 404 pages, especially when google explicitly says that it's a bad practice. 

https://support.google.com/webmasters/answer/181708?hl=en

Why can't we return a headers object from getStaticProps or have some other solution? Redirects would be helpful too.

**Reply by lukaszsobek** [reply]
If I understand correctly the logic seems to be something along the lines:
1. `getStaticPaths` defines whether next should consider other pages than those defined in the `paths` prop using `fallback`.
2. if `fallback` is false everything not matching will be 404'd without even calling the component.
3. if `fallback` is true or we have a match in `paths` we move to getStaticProps trying construct properties for the component to be rendered (we're still on the server)
4. since `getStaticPaths` and `getStaticProps` are async functions the component will have already received the information if it is a fallback page or should 404 by default.
5. we're now at the stage where our code determines what should be rendered.
6. since `getStaticProps` does not render itself, but only provides props to the component to be rendered we decide what the final component looks like.
7. the component gets the props (we're still on the server) and "decides" what to SSG.
8. if the component gets a prop we use as catch for errors e.g. "hasError" it will only render that part of the component to the frontend (we're still on the server).
9. we can decide if we want to make this page 404 or anything else.
10. the result of the SSG is sent to the frontend, whether error page, results or anything else together with the appropriate headers (if you set them that is).

tl;dr - when the server knows it's a 404 it will 404 immediately, in other cases it relies on us to tell it what it should do.

Thoughts:
Maybe we should have a way to short-cuircuit `getStaticprops` to 404 to the default error page immediately (like `getStaticPaths` does when `fallback` is false) without having to go through the whole ordeal of passing props through to the component.

But then again, what if you don't want to 404 but redirect to the homepage? That would take us to a point where we'd probably have to declare there is an error and tell `getStaticProps` to ignore the component altogether and hijack the server response to do what we want. Kinda feel that might be too much for the scope of a function called `getStaticProps`.

On the other hand we could have `getStaticProps` invoke anther function like "onStaticPropsError" if `getStaticProps` returns e.g. `return { props: {...}, error: true};` and let that handle the error on page level.

I like the idea of `getStaticProps` handling errors before the render, just not sure about the returning a headers object or redirecting inside of `getStaticProps` directly part.

**Reply by thadeucity** [reply]
I just posted an article on Medium with my solution for this issue, returning the status code as 404 not 200 when redirected from `getStaticProps()`

https://victorthadeu-66076.medium.com/how-to-return-404-while-using-fallback-blocking-in-next-js-6fda63aad818

**Reply by aguilarm** [reply]
Noting that in Next v10 you can actually return a 404 now https://nextjs.org/blog/next-10#notfound-support it seems.

**Comment by sambokai** [comment]
@timneutkens 

>404 is a state of the UI

I don't think that's the best way to think about this, due to what @stevez86 has said. A 404 is a standardised indication by the server, that the requested resource is missing.

**Comment by JonCognioDigital** [comment]
I'm disappointed (and kind of amazed) that Next would take this route deliberately.

The whole point of using a n SSR framework like Next JS is SEO, I can't think of many other reasons people would need SSR. So, if Next can't be SEO friendly then we may as well just build using create-react-app, leave the server out of it and get a faster app.

Returning a 200 status code for a 404 error is just not right, and how would it be possible to implement a 301 solution for moved pages when using fallback? Seems like this just hasn't been thought through.

**Reply by morganfeeney** [reply]
very good point re: SEO, the 404 status code is always used by Google search console as an indication that errors are being handled correctly.

**Comment by BernardA** [comment]
@timneutkens, in your solution you are assuming that having no posts means a 404 page, and that could be correct in that specific case of a route like `pages/blog/[slug]`. 

More broadly though, having an empty response may also mean that there are currently no items that correspond to that route and then the UI should inform the user about it. No 404 pages or even errors are applicable. 

I am building an app with small ads for multiple categories ( about 200 ) and cities ( could be north of 20, 000 ). 
There is an overall `/search` page and from there one can drill down to `/search/city/rootCategory/parentCategory/category`.

As you can understand it is not realistic to pre-build all pages and there  are a number of ways an user can simply type a completely wrong URL.

So, I was thinking about using `getStaticProps` together with `getStaticPaths` set to `fallback: true`.
I would pre-build the most probable routes and the ones for which I already have ads and leave the others to be handled by `fallback`. 

The question is how to handle the mistyped URLs. I was thinking about doing a sanity check on `getStaticProps` to check if the URL matches some basic criteria, like the format and some of its content.

If the URL does not pass that basic sanity check it would mean it is plainly wrong and I would like to throw a 404 directly from within `getStaticProps`, before attempting to fetch remote data (which would be a waste of time in that case).

Can that be accomplished correctly, meaning with an actual 404 page with the correct HTTP code?

**Comment by beppek** [comment]
The problem with the suggested solution is that it doesn't allow you to build the error page statically as the component is imported rather than the error page used.

Example: building a site with a headless CMS that creates all pages on dynamic routes. These are fetched in `getStaticPaths` and built in `getStaticProps`. For any missing pages there is a custom 404-page content created in the CMS which should be fetched and built statically as well.

With `fallback: false`, `getStaticProps` works as expected and the custom 404 page is built with content from the CMS statically. However, with `fallback: true`, no true 404 is thrown and the above solution would require a second call to the CMS API to fetch the 404-page content.

**Comment by tafelito** [comment]
would like to know more about this but seems that #15672 is the answer for now

**Reply by samstr** [reply]
> would like to know more about this but seems that #15672 is the answer for now

Does `fallback: unstable_blocking` actually deliver 404s though?

I'm doing the following in one of my page components and statusCode 404 doesn't seem to be present in the response at all

``` 
if (!post) {
    return (
      <MainLayout>
        <Error statusCode={404} title="Post not found" />
      </MainLayout>
    );
  }
```

**Reply by jnv** [reply]
I think it doesn't, my understanding is that it's a prerequisite. From discussion in #14890 I assume that there will be an RFC for 404s eventually and it will be handled declaratively in similar vein like redirects are drafted.

**Comment by Octowl** [comment]
My concern to this approach is that very time a page that genuinely doesn't exist is encountered, the fallback code in `getStaticProps` runs, the page page component receives props and some logic detects that the data is missing and renders an error component.

But that means that every time a user lands on a missing page, every time there's a typo in the url, or even through sheer malice, the SSG process will actually run. HTML and JSON files will actually get generated. For every missing page.

![Screen Shot 2020-09-14 at 9 16 32 PM](https://user-images.githubusercontent.com/1828497/93122671-982d7300-f6cf-11ea-97a8-bcef4015178e.png)

This seems wasteful to me and I feel there's definitely a place for throwing or redirecting to a `404` page from `getStaticProps` directly and short-circuiting/skipping the page render.

It also seems dangerous. One tiny script that makes repeated requests to random garbage urls on my site could result in mountains of junk html and json files on the server.

**Reply by Emiliano-Bucci** [reply]
Also i've noticed that when `isFallback` is true this is actually running on the client side. This means that, as the Tim example above, we set `noindex` and `nofollow` in the html tag header, we're actually setting it on the client; if we inspect the page source, we'll see that there (and so what technically the spiders will see) is the orginal setted value.

**Reply by canali83** [reply]
any clear solution? even you return 404 page. nextjs creates an empty file for every request. It's unacceptable

**Reply by fmnxl** [reply]
We need a way to _remove_ the cached page. I'm thinking of allowing `getStaticProps` to return `invalidate: true` to do this.

```
export async function getStaticProps({ params }) {
  try {
    const data = await fetchData(...)
  } catch (e) {
    if (e instance of NotFoundError) { // Some made up exception type
      return { invalidate: true }
    }
  }
}
```

This should redirect to 404. This would fix the issue of both `fallback: true` for "new" routes and also cached routes..

**Comment by Vince-vegas** [comment]
why this code is not prerender in the page
Is this will affect SEO

```
export async function getStaticPaths(){
 paths: [],
 fallback: true
}
```

**Reply by alevosia** [reply]
This is using Incremental Static Generation.

**Comment by AubreyHewes** [comment]
Shall we all get back to the actual problem?

> https://github.com/vercel/next.js/discussions/10960#discussioncomment-16488

i.e. it is not normal for a server to give a 200 on an actual 404.. that the ui says 404 is irrelevant.

**Reply by procarrera** [reply]
agree

**Reply by fmnxl** [reply]
I’m not sure where “404 is just a state of the UI” thought is coming from.

I also think this is an example of a framework trying to take on too many functionalities. There are already relatively more rigorous ways to cache responses e.g. with CDNs and returning `Cache-Control` and `Surrogate-Control` headers, which lets me keep full SSR while getting that static html low latency and reduced server load.

**Comment by anthwinter** [comment]
Just leaving my solution here that throws a 404 page. It might be a bit specific, but hope it helps someone.

I have a Next.js site (v9.5) that brings in data from a headless CMS. You can set the URL of a page in the CMS to be whatever you want.

In my `pages` directory I have `[[...slug]].tsx` that uses getStaticPaths (with revalidate) and getStaticProps.

I use `fallback: false` in getStaticPaths so that it correctly 404's (not 200) and uses my `404.tsx`, page.

My `404.tsx` page uses getStaticProps (that pulls the error page also from the CMS) and also has `revalidate` set. With revalidate set, any new pages that once 404'd, will end up being available.

**Reply by tpperez** [reply]
@anthwinter , it works! It's a very nice solution!

With revalidate at **[[...slug]].tsx**, it will be looking for update at existing pages.
If a new page was created, it will need **fallback: true** to look for updates, if it's not it will be considered as a 404.

At your solution, you inverted the logic: you consider all paths that are not present at the build with **404 status**.
And at **404 page with revalidate**, you will be looking for updates to reconsider a previous 404 status. The page will revalidate and change the status when a new path for this request is created.

Thank you so much!

**Reply by LucasReinaldo** [reply]
> Just leaving my solution here that throws a 404 page. It might be a bit specific, but hope it helps someone.
> 
> I have a Next.js site (v9.5) that brings in data from a headless CMS. You can set the URL of a page in the CMS to be whatever you want.
> 
> In my `pages` directory I have `[[...slug]].tsx` that uses getStaticPaths (with revalidate) and getStaticProps.
> 
> I use `fallback: false` in getStaticPaths so that it correctly 404's (not 200) and uses my `404.tsx`, page.
> 
> My `404.tsx` page uses getStaticPaths (that pulls the error page also from the CMS) and also has `revalidate` set. With revalidate set, any new pages that once 404'd, will end up being available.

Hi mate @anthwinter would you be able to provide an example? I tried to replicate, but when I build it says that 
`getStaticPaths can only be used with dynamic pages, not '/404'.`

Thanks!

**Reply by anthwinter** [reply]
Yeah, sure. I'll post it very soon. However, I made a typo about the 404.tsx in that I use only getStaticProps, not getStaticPaths.

**Reply by anthwinter** [reply]
@LucasReinaldo I've created a gist with the exported functions in. Hope it helps. https://gist.github.com/anthwinter/6c1d6d2d381ac1139f6fe24f5e7e70a5

**Comment by JuanmaMenendez** [comment]
Guys now in version 10, inside **getStaticProps**  you can return "**notFound: true**" to return a 404 status and reedirect the user to the 404.js page or _error.js page. It works even with **fallback: true** set.

https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation

**Reply by anthwinter** [reply]
Oh, nice. I missed the `notFound` property in the new release. Thanks for pointing it out!

**Reply by gavrichards** [reply]
Does this work for getServerSideProps too?

**Reply by IskanderMustafin** [reply]
> Does this work for getServerSideProps too?

It should work according to https://github.com/vercel/next.js/pull/18241

**Reply by jdtzmn** [reply]
@JuanmaMenendez Do you know if I should use `notFound: true` and `fallback: true` or `notFound: true` and `fallback: "blocking"`? Does it matter?

**Reply by alexandersch** [reply]
This does not return a 404 status for me. Just a 200 with notFound set to true in pageProps. I am using getStaticProps with fallback: 'blocking'.

**Reply by JuanmaMenendez** [reply]
> @JuanmaMenendez Do you know if I should use `notFound: true` and `fallback: true` or `notFound: true` and `fallback: "blocking"`? Does it matter?

`notFound: true` should work with both, `fallback: "true"` and `fallback: "blocking" ` in the same way.

**Reply by thadeucity** [reply]
@alexandersch Maybe this article can help you https://victorthadeu-66076.medium.com/how-to-return-404-status-code-while-using-fallback-blocking-in-next-js-ff99f67fc392

If you are using fallback: true, it will always return 200 because the search will occur after the initial page render

**Reply by Fi1osof** [reply]
> Guys now in version 10, inside **getStaticProps** you can return "**notFound: true**" to return a 404 status and reedirect the user to the 404.js page or _error.js page. It works even with **fallback: true** set.
> 
> https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation

But this not works with getInitialProps (dynamic data);

My solution: add logic in _App component for handle statusCode from pages.
Now i can do like this:
```ts
SomePage.getInitialProps = () => {
  // ... load and check data

  return {
    // Return error if data did not received
    statusCode: 404,
  }
}
```

App logic:
here https://github.com/prisma-cms/nextjs/blob/da3c7569c12d2c5fcb526b89b573e470188b9a62/src/pages/_App/index.tsx#L109-L116
and here https://github.com/prisma-cms/nextjs/blob/da3c7569c12d2c5fcb526b89b573e470188b9a62/src/pages/_App/index.tsx#L30-L48

This works on both server and client side.

**Reply by robbertvancaem** [reply]
> Oh, nice. I missed the `notFound` property in the new release. Thanks for pointing it out!

Same here, nice addition tho!

**Reply by Fi1osof** [reply]
> > Oh, nice. I missed the `notFound` property in the new release. Thanks for pointing it out!
> 
> Same here, nice addition tho!

So this will be added soon? :)

**Reply by robbertvancaem** [reply]
@Fi1osof The `notFound` flag is already available since version 10, check out the docs: https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation

If the boolean is `true`, it renders your `404.js` page and runs `getStaticProps` for it as well, making it possible to use CMS-generated content in that page 🎉

**Reply by Fi1osof** [reply]
> @Fi1osof The `notFound` flag is already available since version 10, check out the docs: https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation
> 
> If the boolean is `true`, it renders your `404.js` page and runs `getStaticProps` for it as well, making it possible to use CMS-generated content in that page

I seen that. But it does not works with getInitialProps. Will be good add posibility to return {notFound: true} in getInitialProps() that next App render 404 instead page.

**Reply by pkellner** [reply]
I'm finding the 404 generated when using the notFound flag only returns from JavaScript running in the browser.  That is, if I turn JavaScript off, I get a 304 (not a 404 as I'd expect), and the returned HTML is this.

https://gist.github.com/pkellner/c9959baec4765c792d8b2a93583421da

In dev mode (npm run dev), the 404 comes back as I'd expect, but when I run for prod (npm run build;npm start), I just get the 304.

**Comment by isggwp** [comment]
we can use `fallback: "blocking"` mode on  [**nextjs 10 version**](https://nextjs.org/blog/next-10) .
just passed **notFound: true** when failed inside getStaticProps

inside **getStaticProps**

```
export const getStaticProps = async ({ params }) => {
  try {
    //fetcher
    return {props: {data}}
  } catch (error) {
    return {
      notFound: true
    }
  }
};

```

**Reply by threehams** [reply]
We use `NotFoundError` with an optional message. A great side effect throwing errors is that, in development, you get the message and stack trace so you can track it down if needed; in production, it just looks like any other 404.

**Reply by florianwalther-private** [reply]
> We use `NotFoundError` with an optional message. A great side effect throwing errors is that, in development, you get the message and stack trace so you can track it down if needed; in production, it just looks like any other 404.

So your server throws a 404 and your Next app checks for `error instanceof NotFoundError`? Or is the process different?

**Comment by dgrcode** [comment]
This discussion seems to have an obsolete answer. There's a better way of doing this after some changes based on discussion https://github.com/vercel/next.js/discussions/14890.

Example code form that discussion:
```js
export default function HomePage() {
  return Hello World
}

export async function getStaticProps({params}) {
  const item = await getData(params.id)
  if(!item) {
    return {
      redirect: {
        permanent: false,
        destination: '/somewhere-else'
      }
    }
  }
  
  return {
    props: {
      item
    }
  }
}
```

**Reply by delbaoliveira** [reply]
Thank you! We've updated the answer recommending `notFound: true`.

**Reply by ashishkarki** [reply]
@delbaoliveira `notFound: true` it seems requires `fallBack: blocking` in order to return a 404 status, at least according to my experimentation. A previous [comment](https://github.com/vercel/next.js/discussions/10960#discussioncomment-123885) by  [thadeucity](https://github.com/thadeucity) also says the same thing. Could you please chime in ?
